agent/consul/streaming_endpoint.go

```diff
@@ -65,16 +71,67 @@ func (h *ConsulGRPCAdapter) Subscribe(req *stream.SubscribeRequest, server strea
 		case <-server.Context().Done():
 			return nil
 		case event, ok := <-eventCh:
+			// If the channel was closed, that means the state store filled it up
+			// faster than we could pull events out.
 			if !ok {
 				return fmt.Errorf("handler could not keep up with events")
 			}
-			event.SetACLRules()
-			if !filt.allowEvent(event) {
-				continue
-			}
-			if err := server.Send(&event); err != nil {
+
+			if err := sendEvent(event, eventFilter, aclFilter, eval, sent, server); err != nil {
 				return err
 			}
 		}
 	}
 }
+
+// sendEvent sends the given event along the stream if it passes ACL, boolean, and
+// any topic-specific filtering.
+func sendEvent(event stream.Event, eventFilter stream.EventFilterFunc, aclFilter *aclFilter,
+	eval *bexpr.Evaluator, sent map[string]struct{}, server stream.Consul_SubscribeServer) error {
+	allowEvent := true
+
+	// Check if the event should be filtered based on the request type.
+	if eventFilter != nil && !eventFilter(event) {
+		allowEvent = false
+	}
+
+	// Filter by ACL rules.
+	if !aclFilter.allowEvent(event) {
+		allowEvent = false
+	}
+
+	// Apply the user's boolean expression filtering.
+	if eval != nil {
+		allow, err := eval.Evaluate(event.FilterObject())
+		if err != nil {
+			return err
+		}
+		if !allow {
+			allowEvent = false
+		}
+	}
+
+	// If the event would be filtered but the agent needs to know about
+	// the delete, change the operation to delete and send the event
+	// before removing the ID from the sent map.
+	id := event.ID()
+	if !allowEvent {
+		if _, ok := sent[id]; ok {
+			event.Op = stream.Operation_Delete
+			delete(sent, id)
+		} else {
+			// If the event should be filtered and the agent doesn't know about it,
+			// just return early and don't send anything.
+			return nil
+		}
+	} else {
+		sent[id] = struct{}{}
```

Comment:
@kyhavlov What do you think about a simple 32 bit hash for the key for now and a comment about our collision probability tradeoff? 

I'm actually not super sure about the performance difference I thought int maps wee super optimized but a quick playground test shows them actually using more memory than strings so might be premature.

The one thing I do think we would change is the confusing naming of `ID` here? Can we make it clearer that it is the ID of the _subject_ of that event not a unique identifier for the actual specific event? This is the same thing Kafka calls a "key" but that might be confusing since "key" is something else in our topic model. How about `EntityID` and in doc comments explain what that means and why it's not a unique ID per event?

Reply:
Ah realised why it wasn't working. Int32 would be a decent saving: https://play.golang.org/p/QE_m5UAI8S0

String Map: 637 kb (10000 items)
Int Map: 96 kb (10000 items)

Reply:
This was done in the next PR: https://github.com/hashicorp/consul/pull/6584#pullrequestreview-303998778